File: 2024/day24/zig/solution.zig
const std = @import("std");

const GateOp = enum {
    AND,
    OR,
    XOR,
};

const Gate = struct {
    in1: []const u8,
    in2: []const u8,
    op: GateOp,
    out: []const u8,
};

fn parseInput(allocator: std.mem.Allocator, filename: []const u8) !struct {
    wires: std.StringHashMap(u1),
    gates: std.ArrayList(Gate),
} {
    const file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 10 * 1024 * 1024);
    defer allocator.free(content);

    var wires = std.StringHashMap(u1).init(allocator);
    var gates: std.ArrayList(Gate) = .{};

    var sections = std.mem.splitSequence(u8, content, "\n\n");
    const initial_section = sections.next() orelse return error.InvalidInput;
    const gates_section = sections.next() orelse return error.InvalidInput;

    // Parse initial wire values
    var lines = std.mem.splitScalar(u8, initial_section, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;
        var parts = std.mem.splitSequence(u8, line, ": ");
        const name = parts.next() orelse continue;
        const val_str = parts.next() orelse continue;

        const name_copy = try allocator.dupe(u8, name);
        const val: u1 = if (std.mem.eql(u8, val_str, "1")) 1 else 0;
        try wires.put(name_copy, val);
    }

    // Parse gates
    var gate_lines = std.mem.splitScalar(u8, gates_section, '\n');
    while (gate_lines.next()) |line| {
        if (line.len == 0) continue;

        // Format: "x00 AND y00 -> z00"
        var parts = std.mem.tokenizeScalar(u8, line, ' ');
        const in1 = parts.next() orelse continue;
        const op_str = parts.next() orelse continue;
        const in2 = parts.next() orelse continue;
        _ = parts.next(); // skip "->"
        const out = parts.next() orelse continue;

        const op = if (std.mem.eql(u8, op_str, "AND"))
            GateOp.AND
        else if (std.mem.eql(u8, op_str, "OR"))
            GateOp.OR
        else
            GateOp.XOR;

        try gates.append(allocator, Gate{
            .in1 = try allocator.dupe(u8, in1),
            .in2 = try allocator.dupe(u8, in2),
            .op = op,
            .out = try allocator.dupe(u8, out),
        });
    }

    return .{ .wires = wires, .gates = gates };
}

fn simulate(allocator: std.mem.Allocator, initial_wires: std.StringHashMap(u1), gates: std.ArrayList(Gate)) !std.StringHashMap(u1) {
    var wires = std.StringHashMap(u1).init(allocator);

    // Copy initial wires
    var it = initial_wires.iterator();
    while (it.next()) |entry| {
        const key_copy = try allocator.dupe(u8, entry.key_ptr.*);
        try wires.put(key_copy, entry.value_ptr.*);
    }

    var remaining: std.ArrayList(Gate) = .{};
    defer remaining.deinit(allocator);

    for (gates.items) |gate| {
        try remaining.append(allocator, gate);
    }

    while (remaining.items.len > 0) {
        var made_progress = false;
        var new_remaining: std.ArrayList(Gate) = .{};

        for (remaining.items) |gate| {
            const v1_opt = wires.get(gate.in1);
            const v2_opt = wires.get(gate.in2);

            if (v1_opt != null and v2_opt != null) {
                const v1 = v1_opt.?;
                const v2 = v2_opt.?;

                const result: u1 = switch (gate.op) {
                    .AND => v1 & v2,
                    .OR => v1 | v2,
                    .XOR => v1 ^ v2,
                };

                const out_copy = try allocator.dupe(u8, gate.out);
                try wires.put(out_copy, result);
                made_progress = true;
            } else {
                try new_remaining.append(allocator, gate);
            }
        }

        remaining.deinit(allocator);
        remaining = new_remaining;
    }

    return wires;
}

fn getZValue(wires: std.StringHashMap(u1), allocator: std.mem.Allocator) !u64 {
    var z_wires: std.ArrayList([]const u8) = .{};
    defer z_wires.deinit(allocator);

    var it = wires.iterator();
    while (it.next()) |entry| {
        if (entry.key_ptr.*[0] == 'z') {
            try z_wires.append(allocator, entry.key_ptr.*);
        }
    }

    // Sort in reverse order (z45, z44, ..., z00)
    std.mem.sort([]const u8, z_wires.items, {}, struct {
        fn lessThan(_: void, a: []const u8, b: []const u8) bool {
            return std.mem.order(u8, a, b) == .gt;
        }
    }.lessThan);

    var result: u64 = 0;
    for (z_wires.items) |z| {
        result = (result << 1) | wires.get(z).?;
    }

    return result;
}

fn part1(allocator: std.mem.Allocator, initial_wires: std.StringHashMap(u1), gates: std.ArrayList(Gate)) !u64 {
    var final_wires = try simulate(allocator, initial_wires, gates);
    defer {
        var it = final_wires.keyIterator();
        while (it.next()) |key| {
            allocator.free(key.*);
        }
        final_wires.deinit();
    }

    return try getZValue(final_wires, allocator);
}

fn part2(allocator: std.mem.Allocator, gates: std.ArrayList(Gate)) ![]const u8 {
    var swapped = std.StringHashMap(void).init(allocator);
    defer {
        var it = swapped.keyIterator();
        while (it.next()) |key| {
            allocator.free(key.*);
        }
        swapped.deinit();
    }

    // Build lookup maps
    var gate_by_output = std.StringHashMap(Gate).init(allocator);
    defer gate_by_output.deinit();

    for (gates.items) |gate| {
        try gate_by_output.put(gate.out, gate);
    }

    // Find max bit
    var max_bit: u8 = 0;
    for (gates.items) |gate| {
        if (gate.out[0] == 'z' and gate.out.len == 3) {
            const bit = try std.fmt.parseInt(u8, gate.out[1..], 10);
            if (bit > max_bit) max_bit = bit;
        }
    }

    const max_z = try std.fmt.allocPrint(allocator, "z{d:0>2}", .{max_bit});
    defer allocator.free(max_z);

    for (gates.items) |gate| {
        // Rule: XOR gates that don't take x,y as input should output to z
        if (gate.op == .XOR) {
            const is_xy_xor = (gate.in1[0] == 'x' or gate.in1[0] == 'y') and
                (gate.in2[0] == 'x' or gate.in2[0] == 'y');

            if (!is_xy_xor) {
                // Second-level XOR should output to z
                if (gate.out[0] != 'z') {
                    const out_copy = try allocator.dupe(u8, gate.out);
                    try swapped.put(out_copy, {});
                }
            }
        }

        // Rule: z outputs (except last) should come from XOR
        if (gate.out[0] == 'z' and !std.mem.eql(u8, gate.out, max_z)) {
            if (gate.op != .XOR) {
                const out_copy = try allocator.dupe(u8, gate.out);
                try swapped.put(out_copy, {});
            }
        }

        // Rule: AND gates (except x00 AND y00) should feed into OR
        if (gate.op == .AND) {
            const is_first_bit = (std.mem.eql(u8, gate.in1, "x00") and std.mem.eql(u8, gate.in2, "y00")) or
                (std.mem.eql(u8, gate.in1, "y00") and std.mem.eql(u8, gate.in2, "x00"));

            if (!is_first_bit) {
                var used_by_or = false;
                for (gates.items) |g2| {
                    if (g2.op == .OR) {
                        if (std.mem.eql(u8, g2.in1, gate.out) or std.mem.eql(u8, g2.in2, gate.out)) {
                            used_by_or = true;
                            break;
                        }
                    }
                }
                if (!used_by_or) {
                    const out_copy = try allocator.dupe(u8, gate.out);
                    try swapped.put(out_copy, {});
                }
            }
        }

        // Rule: XOR of x,y should feed into XOR and AND
        if (gate.op == .XOR) {
            const is_xy_xor = (gate.in1[0] == 'x' or gate.in1[0] == 'y') and
                (gate.in2[0] == 'x' or gate.in2[0] == 'y');
            const is_z00 = (std.mem.eql(u8, gate.in1, "x00") and std.mem.eql(u8, gate.in2, "y00")) or
                (std.mem.eql(u8, gate.in1, "y00") and std.mem.eql(u8, gate.in2, "x00"));

            if (is_xy_xor and !is_z00) {
                var used_by_xor = false;
                var used_by_and = false;
                for (gates.items) |g2| {
                    if (std.mem.eql(u8, g2.in1, gate.out) or std.mem.eql(u8, g2.in2, gate.out)) {
                        if (g2.op == .XOR) used_by_xor = true;
                        if (g2.op == .AND) used_by_and = true;
                    }
                }
                if (!used_by_xor or !used_by_and) {
                    const out_copy = try allocator.dupe(u8, gate.out);
                    try swapped.put(out_copy, {});
                }
            }
        }
    }

    // Collect and sort swapped wires
    var swapped_list: std.ArrayList([]const u8) = .{};
    defer swapped_list.deinit(allocator);

    var it = swapped.keyIterator();
    while (it.next()) |key| {
        try swapped_list.append(allocator, key.*);
    }

    std.mem.sort([]const u8, swapped_list.items, {}, struct {
        fn lessThan(_: void, a: []const u8, b: []const u8) bool {
            return std.mem.order(u8, a, b) == .lt;
        }
    }.lessThan);

    // Join with commas
    var result: std.ArrayList(u8) = .{};
    defer result.deinit(allocator);

    for (swapped_list.items, 0..) |wire, i| {
        if (i > 0) try result.append(allocator, ',');
        try result.appendSlice(allocator, wire);
    }

    return try result.toOwnedSlice(allocator);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const parsed = try parseInput(allocator, "../input.txt");
    var wires = parsed.wires;
    defer {
        var it = wires.keyIterator();
        while (it.next()) |key| {
            allocator.free(key.*);
        }
        wires.deinit();
    }

    var gates = parsed.gates;
    defer {
        for (gates.items) |gate| {
            allocator.free(gate.in1);
            allocator.free(gate.in2);
            allocator.free(gate.out);
        }
        gates.deinit(allocator);
    }

    const p1 = try part1(allocator, wires, gates);
    std.debug.print("Part 1: {d}\n", .{p1});

    const p2 = try part2(allocator, gates);
    defer allocator.free(p2);
    std.debug.print("Part 2: {s}\n", .{p2});
}

File: 2024/day24/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAX_WIRES 1000
#define MAX_GATES 500
#define MAX_NAME 4

typedef enum {
    AND,
    OR,
    XOR
} Operation;

typedef struct {
    char in1[MAX_NAME];
    char in2[MAX_NAME];
    Operation op;
    char out[MAX_NAME];
} Gate;

typedef struct {
    char name[MAX_NAME];
    int value;
    bool has_value;
} Wire;

Wire wires[MAX_WIRES];
int wire_count = 0;
Gate gates[MAX_GATES];
int gate_count = 0;

int find_wire(const char *name) {
    for (int i = 0; i < wire_count; i++) {
        if (strcmp(wires[i].name, name) == 0) {
            return i;
        }
    }
    return -1;
}

int get_or_create_wire(const char *name) {
    int idx = find_wire(name);
    if (idx >= 0) return idx;

    if (wire_count >= MAX_WIRES) {
        fprintf(stderr, "Error: Too many wires (max %d)\n", MAX_WIRES);
        exit(1);
    }

    strncpy(wires[wire_count].name, name, MAX_NAME - 1);
    wires[wire_count].name[MAX_NAME - 1] = '\0';
    wires[wire_count].value = 0;
    wires[wire_count].has_value = false;
    return wire_count++;
}

void parse_input(const char *filename) {
    FILE *f = fopen(filename, "r");
    if (!f) {
        perror("Failed to open input file");
        exit(1);
    }

    char line[256];
    bool reading_gates = false;

    while (fgets(line, sizeof(line), f)) {
        // Remove newline
        line[strcspn(line, "\n")] = 0;

        if (strlen(line) == 0) {
            reading_gates = true;
            continue;
        }

        if (!reading_gates) {
            // Parse initial wire values: "x00: 1"
            char name[MAX_NAME];
            int value;
            sscanf(line, "%3s: %d", name, &value);
            int idx = get_or_create_wire(name);
            wires[idx].value = value;
            wires[idx].has_value = true;
        } else {
            // Parse gates: "x00 AND y00 -> z00"
            char in1[MAX_NAME], op_str[4], in2[MAX_NAME], out[MAX_NAME];
            sscanf(line, "%3s %3s %3s -> %3s", in1, op_str, in2, out);

            if (gate_count >= MAX_GATES) {
                fprintf(stderr, "Error: Too many gates (max %d)\n", MAX_GATES);
                exit(1);
            }

            strncpy(gates[gate_count].in1, in1, MAX_NAME - 1);
            gates[gate_count].in1[MAX_NAME - 1] = '\0';
            strncpy(gates[gate_count].in2, in2, MAX_NAME - 1);
            gates[gate_count].in2[MAX_NAME - 1] = '\0';
            strncpy(gates[gate_count].out, out, MAX_NAME - 1);
            gates[gate_count].out[MAX_NAME - 1] = '\0';

            if (strcmp(op_str, "AND") == 0) {
                gates[gate_count].op = AND;
            } else if (strcmp(op_str, "OR") == 0) {
                gates[gate_count].op = OR;
            } else {
                gates[gate_count].op = XOR;
            }

            // Ensure all wires exist
            get_or_create_wire(in1);
            get_or_create_wire(in2);
            get_or_create_wire(out);

            gate_count++;
        }
    }

    fclose(f);
}

void simulate() {
    bool made_progress;
    do {
        made_progress = false;

        for (int i = 0; i < gate_count; i++) {
            int in1_idx = find_wire(gates[i].in1);
            int in2_idx = find_wire(gates[i].in2);
            int out_idx = find_wire(gates[i].out);

            if (wires[in1_idx].has_value && wires[in2_idx].has_value && !wires[out_idx].has_value) {
                int v1 = wires[in1_idx].value;
                int v2 = wires[in2_idx].value;
                int result;

                switch (gates[i].op) {
                    case AND:
                        result = v1 & v2;
                        break;
                    case OR:
                        result = v1 | v2;
                        break;
                    case XOR:
                        result = v1 ^ v2;
                        break;
                }

                wires[out_idx].value = result;
                wires[out_idx].has_value = true;
                made_progress = true;
            }
        }
    } while (made_progress);
}

long long get_z_value() {
    // Collect all z wires
    char z_wires[100][MAX_NAME];
    int z_count = 0;

    for (int i = 0; i < wire_count; i++) {
        if (wires[i].name[0] == 'z') {
            strcpy(z_wires[z_count++], wires[i].name);
        }
    }

    // Sort z wires in descending order (z45, z44, ..., z00)
    for (int i = 0; i < z_count - 1; i++) {
        for (int j = i + 1; j < z_count; j++) {
            if (strcmp(z_wires[i], z_wires[j]) < 0) {
                char temp[MAX_NAME];
                strcpy(temp, z_wires[i]);
                strcpy(z_wires[i], z_wires[j]);
                strcpy(z_wires[j], temp);
            }
        }
    }

    // Build the number
    long long result = 0;
    for (int i = 0; i < z_count; i++) {
        int idx = find_wire(z_wires[i]);
        result = (result << 1) | wires[idx].value;
    }

    return result;
}

int find_gate_by_output(const char *out) {
    for (int i = 0; i < gate_count; i++) {
        if (strcmp(gates[i].out, out) == 0) {
            return i;
        }
    }
    return -1;
}

int find_gate_by_inputs_op(const char *a, const char *b, Operation op) {
    for (int i = 0; i < gate_count; i++) {
        if (gates[i].op == op) {
            if ((strcmp(gates[i].in1, a) == 0 && strcmp(gates[i].in2, b) == 0) ||
                (strcmp(gates[i].in1, b) == 0 && strcmp(gates[i].in2, a) == 0)) {
                return i;
            }
        }
    }
    return -1;
}

bool is_input_to_gate_with_op(const char *wire, Operation op) {
    for (int i = 0; i < gate_count; i++) {
        if (gates[i].op == op) {
            if (strcmp(gates[i].in1, wire) == 0 || strcmp(gates[i].in2, wire) == 0) {
                return true;
            }
        }
    }
    return false;
}

bool starts_with(const char *str, const char *prefix) {
    return strncmp(str, prefix, strlen(prefix)) == 0;
}

char *part2() {
    static char result[256];
    char swapped[32][MAX_NAME];  // Increased to handle more potential swaps
    int swap_count = 0;

    // Find the highest z bit
    int max_bit = 0;
    for (int i = 0; i < gate_count; i++) {
        if (gates[i].out[0] == 'z') {
            int bit = atoi(gates[i].out + 1);
            if (bit > max_bit) max_bit = bit;
        }
    }

    char max_z[MAX_NAME];
    sprintf(max_z, "z%02d", max_bit);

    for (int i = 0; i < gate_count; i++) {
        Gate *g = &gates[i];

        // Rule: XOR gates that don't take x,y as input should output to z
        if (g->op == XOR) {
            bool is_xy_xor = (starts_with(g->in1, "x") || starts_with(g->in1, "y")) &&
                            (starts_with(g->in2, "x") || starts_with(g->in2, "y"));

            if (!is_xy_xor) {
                // This is a second-level XOR, should output to z
                if (!starts_with(g->out, "z")) {
                    if (swap_count < 32) {
                        strcpy(swapped[swap_count++], g->out);
                    }
                }
            }
        }

        // Rule: z outputs (except final carry) should come from XOR
        if (starts_with(g->out, "z") && strcmp(g->out, max_z) != 0) {
            if (g->op != XOR) {
                if (swap_count < 32) {
                    strcpy(swapped[swap_count++], g->out);
                }
            }
        }

        // Rule: AND gates (except x00 AND y00) should feed into OR
        if (g->op == AND) {
            bool is_first_bit = (strcmp(g->in1, "x00") == 0 && strcmp(g->in2, "y00") == 0) ||
                               (strcmp(g->in1, "y00") == 0 && strcmp(g->in2, "x00") == 0);

            if (!is_first_bit) {
                // This AND output should be input to an OR gate
                if (!is_input_to_gate_with_op(g->out, OR)) {
                    if (swap_count < 32) {
                        strcpy(swapped[swap_count++], g->out);
                    }
                }
            }
        }

        // Rule: XOR of x,y should feed into XOR and AND
        if (g->op == XOR) {
            bool is_xy_xor = (starts_with(g->in1, "x") || starts_with(g->in1, "y")) &&
                            (starts_with(g->in2, "x") || starts_with(g->in2, "y"));

            bool is_z00 = (strcmp(g->in1, "x00") == 0 && strcmp(g->in2, "y00") == 0) ||
                         (strcmp(g->in1, "y00") == 0 && strcmp(g->in2, "x00") == 0);

            if (is_xy_xor && !is_z00) {
                bool used_by_xor = is_input_to_gate_with_op(g->out, XOR);
                bool used_by_and = is_input_to_gate_with_op(g->out, AND);

                if (!(used_by_xor && used_by_and)) {
                    if (swap_count < 32) {
                        strcpy(swapped[swap_count++], g->out);
                    }
                }
            }
        }
    }

    // Remove duplicates
    for (int i = 0; i < swap_count - 1; i++) {
        for (int j = i + 1; j < swap_count; j++) {
            if (strcmp(swapped[i], swapped[j]) == 0) {
                // Remove duplicate
                for (int k = j; k < swap_count - 1; k++) {
                    strcpy(swapped[k], swapped[k + 1]);
                }
                swap_count--;
                j--;
            }
        }
    }

    // Sort alphabetically
    for (int i = 0; i < swap_count - 1; i++) {
        for (int j = i + 1; j < swap_count; j++) {
            if (strcmp(swapped[i], swapped[j]) > 0) {
                char temp[MAX_NAME];
                strcpy(temp, swapped[i]);
                strcpy(swapped[i], swapped[j]);
                strcpy(swapped[j], temp);
            }
        }
    }

    // Build comma-separated result
    result[0] = '\0';
    for (int i = 0; i < swap_count; i++) {
        if (i > 0) strcat(result, ",");
        strcat(result, swapped[i]);
    }

    return result;
}

int main() {
    parse_input("../input.txt");

    // Part 1: Simulate the circuit
    simulate();
    long long z_value = get_z_value();
    printf("Part 1: %lld\n", z_value);

    // Part 2: Find swapped wires
    char *swapped_wires = part2();
    printf("Part 2: %s\n", swapped_wires);

    return 0;
}
